internal/controller/maintenancereplica_controller.go

```diff
@@ -0,0 +1,217 @@
+	log := log.FromContext(ctx)
+	if !replica.DeletionTimestamp.IsZero() {
+		return r.delete(ctx, log, replica)
+	}
+
+	if modified, err := clientutils.PatchEnsureFinalizer(ctx, r.Client, replica, ServerMaintenanceReplicaFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Ensured finalizer has been added")
+
+	return r.reconcile(ctx, log, replica)
+}
+
+func (r *ServerMaintenanceReplicaReconciler) delete(ctx context.Context, log logr.Logger, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log.V(1).Info("Deleting ServerMaintenanceReplica")
+
+	log.V(1).Info("Ensuring that the finalizer is removed")
+	if modified, err := clientutils.PatchEnsureNoFinalizer(ctx, r.Client, replica, ServerMaintenanceReplicaFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Deleted ServerMaintenanceReplica")
+	return ctrl.Result{}, nil
+}
+
+func (r *ServerMaintenanceReplicaReconciler) reconcile(ctx context.Context, log logr.Logger, replica *metalv1alpha1.ServerMaintenanceReplica) (ctrl.Result, error) {
+	log.V(1).Info("Reconciling ServerMaintenanceReplica")
+	servers, err := r.getServersBySelector(ctx, replica)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(servers.Items) == 0 {
+		log.V(1).Info("No servers found")
+		return ctrl.Result{}, nil
+	}
+
+	maintenancelist := &metalv1alpha1.ServerMaintenanceList{}
+	if err = r.List(ctx, maintenancelist, client.MatchingLabels{
+		ServerMaintenanceReplicaFinalizer: replica.Name,
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If all servers have a maintenance object, only update the status
+	if len(servers.Items) == len(maintenancelist.Items) {
+		if modified, err := r.patchStatus(ctx, replica, calculateStatus(replica, maintenancelist.Items)); err != nil || modified {
+			return ctrl.Result{}, err
+		}
+	}
+
+out:
+	for _, server := range servers.Items {
+		log.V(1).Info("Reconciling server", "server", server.Name)
+		// Check if the server already got a maintenance
+		for _, maintenance := range maintenancelist.Items {
+			if maintenance.Spec.ServerRef.Name == server.Name {
+				log.V(1).Info("Server is already in maintenance", "server", server.Name)
+				continue out
+			}
+		}
+		maintenance := metalv1alpha1.ServerMaintenance{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      replica.Name,
+				Namespace: replica.Namespace,
+				Labels: map[string]string{
+					ServerMaintenanceReplicaFinalizer: replica.Name,
+				},
+			},
+			Spec: metalv1alpha1.ServerMaintenanceSpec{
+				ServerRef:                       v1.LocalObjectReference{Name: server.Name},
+				Policy:                          replica.Spec.Template.Policy,
+				ServerPower:                     replica.Spec.Template.ServerPower,
+				ServerBootConfigurationTemplate: replica.Spec.Template.ServerBootConfigurationTemplate,
+			},
```

Comment:
Should we do:
```go
maintenance.Spec = replica.Spec.Template
maintenance.Spec.ServerRef = v1.LocalObjectReference{Name: server.Name}
```
to be compatible with future updates to `ServerMaintenanceSpec`?